fix array_upper and array_lower for multidimensional arrays

BaurzhanSakhariev opened the pull request:
Resolves https://github.com/crate/crate/issues/12762

proddata commented:
I think this might need another iteration.
CrateDB allows multidimensional arrays with different dimensions for subarrays.

```sql
SELECT [[1,2],[4]] as arr,
````
PostgreSQL does not.
```sql
SELECT '{{1,2},{3}}'::integer[][];
               ^
DETAIL:  Multidimensional arrays must have sub-arrays with matching dimensions.
```
It is not well defined, what array_length returns

BaurzhanSakhariev commented:
> CrateDB allows multidimensional arrays with different dimensions for subarrays.

Yes, I'm currently working on fixing 

```
 @Test
public void testMultiDimensionArrayValidDimension() {
      assertEvaluate("array_upper([[1, 2, 3], [3, 4]], 2)", 2);
}
```
  
In PG14 it's not allowed, so I didn't handle it initially.
  ```
select array_upper(array[array[1, 2, 3], array[3, 4]], 2);
// ERROR:  multidimensional arrays must have array expressions with matching dimensions
```

I will tweak this PR to handle dimension properly and still allow mixed length arrays in order not to make it breaking change.

BaurzhanSakhariev commented:
Hi @mfussenegger, could you please provide an early feedback on `array_upper` update before I change `array_lower`?

BaurzhanSakhariev commented:
unrelated but reproducible failure:
` ./gradlew :server:test --tests "io.crate.integrationtests.TransportSQLActionTest.test_types_with_storage_can_be_inserted_and_queried" -Dtests.seed=7E471C2AE93A688E -Dtests.locale=es-419 -Dtests.timezone=Asia/Aden`

will look int it separately

retest this please

BaurzhanSakhariev commented on docs/general/builtins/scalar-functions.rst:
```diff
@@ -2614,6 +2617,18 @@ Returns: ``integer``
     +------+
     SELECT 1 row in set (... sec)
 
+If there is at least one empty array or ``NULL`` on the requested dimension
```
Not sure - with upper it was clear that we go for max but how we compare NULL and 1? Cannot also align with PG as they don't allow mixing sizes so all sibling arrays on some dimension are same, in our case we have to choose.

mfussenegger replied:
sounds okay to me to do it like this.

BaurzhanSakhariev commented:
Hi @mfussenegger, could you have anotehr look? Apparently, initial approach was wrong, updated test to cover tricky case and implementation in https://github.com/crate/crate/pull/12763/commits/b95e4cc43ce6d6652c5bf67945ce194c3a52cac5
and updated lower in https://github.com/crate/crate/pull/12763/commits/a3f55311d9f8b6d52e408060ec4ae539ad11acbd
